### PR TITLE
Add Router contract and tests

### LINKED ISSUE #21: Write the router contract
Write the router contract

### PR DESCRIPTION
**Pull request by trizin**
Fixes #21

Changes proposed in this PR:

- Add Router contract and tests

### CODE REVIEW

**Comment by alexcos20** [comment]
should we have a function called 'releaseForAll' which calls 'release(IERC20 token, address account)' for all accounts? this will save gas

File: contracts/Router.sol [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    event PaymentReceived(address from, uint256 amount);
+    event PayeeRemoved(address account, uint256 shares);
+    event PayeeShareAdjusted(address account, uint256 shares, uint256 oldShares);
+
+    uint256 private _totalShares;
+    uint256 private _totalReleased;
+
+    mapping(address => uint256) private _shares;
+    mapping(address => uint256) private _released;
+    address[] private _payees;
```
Comment (alexcos20) [inline contracts/Router.sol:42]: should this be private ?

**Comment by jamiehewitt15** [comment]
Why did you copy & paste the code from `PaymentSplitter.sol`? To me it seems like it would be simpler and leave less room for error if we inherited from that contract.

```
import "OpenZeppelin/openzeppelin-contracts@4.7.0/contracts/finance/PaymentSplitter.sol";

contract PaymentSplitter is Context, Ownable, PaymentSplitter {
```

**Comment by jamiehewitt15** [comment]
It would have saved a bit of time if we did this with hardhat as then we could copy the [tests](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/test/finance/PaymentSplitter.test.js) that OpenZeppelin have already written

File: contracts/Router.sol [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        _payees.push(account);
+        _shares[account] = shares_;
+        _totalShares = _totalShares + shares_;
+        emit PayeeAdded(account, shares_);
+    }
+
+    function _removePayee(address account) private {
+        require(account != address(0), "PaymentSplitter: account is the zero address");
+        require(_shares[account] > 0, "PaymentSplitter: account has no shares");
+
```
Comment (jamiehewitt15) [inline contracts/Router.sol:221]: Currently this allows all payees to be removed, I'm not sure if this intentional? It may make sense to have a minimum of one payee

File: contracts/Router.sol [hunk 3]
```diff
@@ -0,0 +1,261 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts (last updated v4.8.0) (finance/PaymentSplitter.sol)
+
+pragma solidity ^0.8.0;
+
+import "OpenZeppelin/openzeppelin-contracts@4.7.0/contracts/token/ERC20/utils/SafeERC20.sol";
+import "OpenZeppelin/openzeppelin-contracts@4.7.0/contracts/utils/Address.sol";
+import "OpenZeppelin/openzeppelin-contracts@4.7.0/contracts/utils/Context.sol";
+import "OpenZeppelin/openzeppelin-contracts@4.7.0/contracts/access/Ownable.sol";
+
```
Comment (jamiehewitt15) [inline contracts/Router.sol:10]: Is there a reason that we're not using `SafeMath.sol`?

File: contracts/Router.sol [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+    function _adjustShare(address account, uint256 shares_) private {
+        require(account != address(0), "PaymentSplitter: account is the zero address");
+        require(shares_ > 0, "PaymentSplitter: shares are 0");
+        require(_shares[account] > 0, "PaymentSplitter: account has no shares");
+
+        uint256 oldShares = _shares[account];
+        _shares[account] = shares_;
+        _totalShares = _totalShares - oldShares + shares_;
+
```
Comment (jamiehewitt15) [inline contracts/Router.sol:245]: We could write this without creating an additional variable which would save a bit of gas:

```suggestion
        _totalShares = _totalShares - _shares[account] + shares_;
        _shares[account] = shares_;

```

File: contracts/Router.sol [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+        _addPayee(account, shares_);
+    }
+
+    function removePayee(address account) external onlyOwner {
+        _removePayee(account);
+    }
+
+    function adjustShare(address account, uint256 shares_) external onlyOwner {
+        _adjustShare(account, shares_);
+    }
```
Comment (jamiehewitt15) [inline contracts/Router.sol:259]: What's the point of these two functions? We can just add the `external onlyOwner` modifiers to the functions above, `_removePayee` and `_adjustShare` are only used here

File: contracts/Router.sol [hunk 6]
```diff
@@ -0,0 +1,261 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts (last updated v4.8.0) (finance/PaymentSplitter.sol)
```
Comment (jamiehewitt15) [inline contracts/Router.sol:2]: This doesn't seem right given that we have modified the contract, also we usually use the Apache-2.0 license. I'd suggest updating to include all contributors:

 ```suggestion
// Copyright OpenZeppelin, BigchainDB GmbH and Ocean Protocol contributors
// SPDX-License-Identifier: (Apache-2.0 AND MIT)
```

Comment (jamiehewitt15) [reply contracts/Router.sol:2]: Although, this isn't needed if we just inherit form `PaymentSplitter.sol` which would be cleaner. Then we should just change the license identifier and copyright info to our own in this file.

Comment (jamiehewitt15) [reply contracts/Router.sol:2]: It would follow the pattern that we have used previously: [example](https://github.com/oceanprotocol/contracts/blob/main/contracts/pools/balancer/BFactory.sol#L2)

**Comment by jamiehewitt15** [comment]
> What's the benefit of rewriting tests when the code being tested remains unchanged?

That was the point I was making - the tests could be duplicated here, which would give us full test coverage without rewriting anything. At the moment we have a large part of this contract that isn't included in our tests.

File: contracts/Router.sol (hunk 1 above)
Comment (alexcos20) [reply contracts/Router.sol:21]: if you know the account(s).  what if you do not ?

**Comment by jamiehewitt15** [comment]
Can you add the NatSpec comments to explain the functions and their variables etc?

File: contracts/Router.sol [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    function adjustShare(address account, uint256 shares_) external onlyOwner {
+        _adjustShare(account, shares_);
+    }
+
+
+    
+    // ---------------------------- private functions ----------------------------
+
+    function _addPayee(address account, uint256 shares_) private {
+        require(account != address(0), "Router: account is the zero address");
```
Comment (jamiehewitt15) [inline contracts/Router.sol:89]: We're not currently testing this require statement fails when the account is the zero address. We should add a test where we attempt to add the zero address as a payee

File: contracts/Router.sol [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+        _payees.push(account);
+        _shares[account] = shares_;
+        _totalShares = _totalShares + shares_;
+        emit PayeeAdded(account, shares_);
+    }
+
+
+    function _removePayee(address account) private {
+        require(account != address(0), "Router: account is the zero address");
```
Comment (jamiehewitt15) [inline contracts/Router.sol:101]: We're not currently testing this require statement fails when the account is the zero address. We should add a test where we attempt to remove the zero address as a payee

File: contracts/Router.sol [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+                emit PayeeRemoved(account, _shares[account]);
+                _shares[account] = 0;
+                _payees.pop();
+                break;
+            }
+        }
+    }
+
+    function _adjustShare(address account, uint256 shares_) private {
+        require(account != address(0), "Router: account is the zero address");
```
Comment (jamiehewitt15) [inline contracts/Router.sol:117]: We're not currently testing this require statement fails when the account is the zero address. We should add a test where we attempt to adjust the shares for the zero address

File: contracts/Router.sol [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+
+    
+    // ---------------------------- private functions ----------------------------
+
+    function _addPayee(address account, uint256 shares_) private {
+        require(account != address(0), "Router: account is the zero address");
+        require(shares_ > 0, "Router: shares are 0");
+        require(_shares[account] == 0, "Router: account already has shares");
```
Comment (jamiehewitt15) [inline contracts/Router.sol:91]: We're not currently testing that this fails if the account already has shares. We should add a test for that

File: contracts/Router.sol [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        _shares[account] = shares_;
+        _totalShares = _totalShares - oldShares + shares_;
+        emit PayeeShareAdjusted(account, shares_, oldShares);
+    }
+
+
+    //---------------------------- fallback ----------------------------
+   
+    receive() external payable virtual {
+        revert("Router: cannot receive ether");
```
Comment (jamiehewitt15) [inline contracts/Router.sol:131]: We haven't covered this in the tests. Let's add a test where we send ether to the contract

File: contracts/Router.sol [hunk 12]
```diff
@@ -0,0 +1,208 @@
+// BigchainDB GmbH and Ocean Protocol contributors
+// SPDX-License-Identifier: (Apache-2.0 AND MIT)
+
+pragma solidity ^0.8.0;
```
Comment (alexcos20) [inline contracts/Router.sol:4]:
```suggestion
pragma solidity ^0.8.19;
```

File: contracts/Router.sol [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+    event PayeeShareAdjusted(address account, uint256 shares, uint256 oldShares);
+
+    // mapping of payee addresses to their shares
+    mapping(address => uint256) private _shares;
+
+    // mapping of payee addresses to a mapping of token addresses to released amount
+    mapping(address => mapping(address => uint256)) private _released;
+
+    // list of payee addresses
+    address[] public _payees;
```
Comment (alexcos20) [inline contracts/Router.sol:38]:
```suggestion
    address[] private _payees;
```

and let's have a getter for it
(sorry for the confusion)

File: contracts/Router.sol [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+        require(payees.length == shares_.length, "Router: payees and shares length mismatch");
+        require(payees.length > 0, "Router: no payees");
+
+        for (uint256 i = 0; i < payees.length; i++) {
+            _addPayee(payees[i], shares_[i]);
+        }
+    }
+
+    // ---------------------------- getters ----------------------------
+
```
Comment (alexcos20) [inline contracts/Router.sol:55]:
```suggestion
    /**
     * @notice Gets payees.
     * @return Array of addresses
     */
    function getPayees() public view returns (addresses[]) {
        return payees;
    }
```

**Review by alexcos20 — CHANGES_REQUESTED** [review]

**Review by alexcos20 — APPROVED** [review]

**Comment by jamiehewitt15** [comment]
I'm getting a compiler error, could you have a look? 

```
$ brownie compile
Brownie v1.19.3 - Python development framework for Ethereum

Compiling contracts...
  Solc version: 0.8.19
  Optimizer: Enabled  Runs: 200
  EVM Version: Istanbul
CompilerError: solc returned the following errors:

DeclarationError: Identifier not found or not unique.
  --> contracts/Router.sol:59:47:
   |
59 |     function getPayees() public view returns (addresses[]) {
   |  
```

**Comment by jamiehewitt15** [comment]
Compiler issue is fixed but I'm getting a failed test now: 

```
$ brownie test
Brownie v1.19.3 - Python development framework for Ethereum

========================================= test session starts =========================================
platform linux -- Python 3.8.9, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /home/jamie/Desktop/ocean/vw-cli
plugins: eth-brownie-1.19.3, forked-1.4.0, hypothesis-6.27.3, xdist-1.34.0, web3-5.31.3
collected 27 items                                                                                    

tests/test_Simpletoken.py ...                                                                   [ 11%]
tests/test_VestingWalletCliff.py ....                                                           [ 25%]
tests/test_VestingWalletHalving.py ...                                                          [ 37%]
tests/test_VestingWalletLinear.py ...                                                           [ 48%]
tests/test_router.py .............F                                                             [100%]

============================================== FAILURES ===============================================
______________________________________ test_send_eth_to_contract ______________________________________

    def test_send_eth_to_contract():
        router = _deployRouter([100], [alice])
        with brownie.reverts():
>           alice.transfer(router, 100)

tests/test_router.py:199: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def test_send_eth_to_contract():
        router = _deployRouter([100], [alice])
        with brownie.reverts():
>           alice.transfer(router, 100)
E           ValueError: insufficient funds for gas * price + value

tests/test_router.py:199: ValueError
========================================== warnings summary ===========================================
tests/test_VestingWalletHalving.py::test_tokenFunding
tests/test_VestingWalletHalving.py::test_tokenFunding
tests/test_VestingWalletHalving.py::test_tokenFunding_big_supply
tests/test_VestingWalletHalving.py::test_tokenFunding_big_supply
  /home/jamie/Desktop/ocean/vw-cli/venv/lib/python3.8/site-packages/eth_abi/codec.py:191: DeprecationWarning: abi.decode_abi() is deprecated and will be removed in version 4.0.0 in favor of abi.decode()
    warnings.warn(

tests/test_VestingWalletHalving.py::test_tokenFunding
tests/test_VestingWalletHalving.py::test_tokenFunding_big_supply
  /home/jamie/Desktop/ocean/vw-cli/venv/lib/python3.8/site-packages/eth_abi/codec.py:164: DeprecationWarning: abi.decode_single() is deprecated and will be removed in version 4.0.0 in favor of abi.decode()
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/warnings.html
======================================= short test summary info =======================================
FAILED tests/test_router.py::test_send_eth_to_contract - ValueError: insufficient funds for gas * pr...
============================== 1 failed, 26 passed, 6 warnings in 10.65s ==============================
Terminating local RPC client...
```

Could you add the test command to the CI so that the tests are also run in github actions before we merge?
